add some scale notes to the limitations page (#42713)
sourcegraph/sourcegraph: doc/code_insights/explanations/current_limitations_of_code_insights.md

* Using a more granular query
* Changing your site configuration so that the timeout is increased, provided your instance setup allows it. [More information on timeouts](https://docs.sourcegraph.com/code_search/how-to/exhaustive#timeouts).

## General scale limitations 

Note: We are working on improvements for the items below in FY23Q4.

> 
> Code Insights is disabled by default on single-docker deployment methods.

Reviewer: ```suggestion

Code Insights is disabled by default on single-docker deployment methods.
```

I think if we want to document this on this page it should be its own header – this is not super related to scale limits IMO (at any scale, this is not supported).

Author: It always seems to be in the same conversation, so I'm going to err on the side of keeping it here in hopes that someone reading about scale will also read about this.

There are a few factors to consider with respect to scale and expected performance.

> 1. General permissiveness - instances that are more open (users can see most repos) will perform better than instances that are more restricted. It is possible to have enough restricted repositories that users cannot render Code Insights.

Reviewer 2: I think this will basically be fixed by https://github.com/sourcegraph/sourcegraph/pull/42516, at least until we start making changes to the series point data structure and may need to adjust. Having more restricted repositories actually ends up speeding things up in some cases because there is less data involved in the aggregation.

2. Number of repositories - Code Insights is well tested to ~35,000 repositories. Users should expect at least linear degredation as repository count grows in both time to calcluate insights, and render performance.

> 3. Large monorepos - Code Insights allocates a fixed amount of time for each query, so large repositories that cause query timeouts will likely not have exhaustive (and therefore accurate results). 

Reviewer: ```suggestion
3. Large monorepos - Code Insights allocates a fixed amount of time for each query, so large repositories that cause query timeouts will likely not have exhaustive (and therefore accurate) results. Until we add more visibility to this state, a heuristic indicator for if this is a problem is seeing values "jump" (either a significant increase or decrease) between the backfilled datapoints on creation and the up-to-date datatpoints added after creation. 
```

4. High cardinaltiy capture groups - When using a capture group insight, high cardinality matches (for example 1000 distinct matches per repository) will cause significant increase in loading times of charts. It is possible to exceed request timeouts if there are too many distinct matches.

> 5. Concurrent users

Reviewer: ```suggestion
5. Concurrent usage
```

>   1. If there are many insight creators the insights will take longer to calculate.

Reviewer: ```suggestion
  1. If there are many insights created at once, the insights will take longer to initially backfill.
```

Author: It's also true that the snapshot and recording times will grow, not just backfills.

>   2. If there are more insight viewers loading times of charts may be impacted.

Reviewer: ```suggestion
  2. If there are very high numbers of insight viewers, loading times of charts may be impacted.
```

Author: I'm not sure we can say that for sure. We don't currently have any sense of where this number falls, and I'd be surprised if it was very high (which to me implies thousands).
